pkg/test/framework/components/bookinfo/bookinfo.go

```diff
@@ -0,0 +1,87 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package bookinfo
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+
+	"istio.io/istio/pkg/test/deployment"
+	"istio.io/istio/pkg/test/framework/dependency"
+	"istio.io/istio/pkg/test/framework/environment"
+	"istio.io/istio/pkg/test/framework/environments/kubernetes"
+)
+
+var (
+
+	// KubeComponent is a component for the Kubernetes environment.
+	KubeComponent = &component{}
+
+	_ io.Closer = &bookInfo{}
+)
+
+type component struct {
+}
+
+type bookInfo struct {
+	deployment *deployment.Instance
+	env        *kubernetes.Implementation
+}
+
+// ID implements implements component.Component.
+func (c *component) ID() dependency.Instance {
+	return dependency.BookInfo
+}
+
+// Requires implements implements component.Component.
+func (c *component) Requires() []dependency.Instance {
+	return []dependency.Instance{
+		dependency.Kubernetes,
+	}
+}
+
+// Init implements implements component.Component.
+func (c *component) Init(ctx environment.ComponentContext, _ map[dependency.Instance]interface{}) (interface{}, error) {
```

Comment:
I've played around with Component approach and there is an impedance mismatch problem. When modeled as a regular dependency, we have two options:
1. It gets reset and redeployed every time to a new test namespace. This is not desirable, as it has a high-reinstallation cost, and not every test may need the dependency.
2. It gets deployed to a dependency namespace once. In this mode, it is not useful to the tests, as the test configuration needs to be applied to the same (i.e. dep namespace as well), which is prone to cross-test interference.

I've found a hybrid approach, where BookInfo is deployed as an opt in by the user (i.e. he has to explicitly call env.DeployBookInfo() in their code). This causes some object model/api warts, but otherwise works.

We can map this to a generic component model internally, but I am not sure how much value we will get out of it. This kind of deployment is likely to be unique, and BookInfo might be the only customer of it. We can still do this, it is a relatively easy change to do, but may not yield too much savings, in terms of infrastructure complexity. Still, it is a more normalized model.

Reply:
Yeah, this sort of thing is something we definitely need to hash out.  This is more common than not, really.  You can think of the pilot test services to be in the same boat, really.

It's fine for now.